zkevm-circuits/src/evm_circuit/execution/address.rs

```diff
@@ -28,7 +32,7 @@ impl<F: Field> ExecutionGadget<F> for AddressGadget<F> {
     const EXECUTION_STATE: ExecutionState = ExecutionState::ADDRESS;
 
     fn configure(cb: &mut EVMConstraintBuilder<F>) -> Self {
-        let address = cb.query_account_address();
```

Comment:
For `address` read part, can use just default word2 which just 2 limbs. Then assign first 16 byte to `lo` limb, and 4 bytes to `hi` limb

Reply:
Yes.  In this case we're doing a read (cb.call_context_lookup_read, CallContextFieldTag::CalleeAddress), so we can just use 2 limbs to get that address and we can assume it's the right type, which means (lo: 16 byte, hi: 4 byte).
Next we have the `cb.stack_push` which uses that address.  Since the address is in a word obtained from a read, we already know it's correct, so there's no need to do any extra range check on this word in this gadget.

The same logic would apply if we read an Address from the RwTable, and we need to built a single expression.  We can just do: `let addr = addr_word.lo() + addr_word.hi() << 2^128`, and this new value is guaranteed to be 160 bits following the previous assumptions.

Reply:
Sorry let me give an update because I think we can't use `cb.query_word_unchecked()` here.
In this opcode we're:
- Reading `address` from CallContext->CalleeAddress
- Writing `address` to Stack

Following the spec, stack values are Words, but CallContext->CalleeAddress is single field value https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/tables.md#rw_table
So we read `[lo: address, hi: 0]`, and we write `[lo: address_lo, hi: address_hi]` where address_lo is the first 128 bits and address_hi is the remaining bits.

We need to guarantee the range check at the write, and since `address_lo` and `address_hi` will be new expressions, we need to range check them.  For example we can split `address` into range-checked bytes, and then from the bytes build the address_lo and address_hi expressions.

Reply:
Thanks for the review and explanation!

I checked [and](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/tables.md#rw_table) and it seems that in CallContext, only CalleeAddress/CallerAddress those 2 currently cause value lo exceeding 128 bit, so it wont be a huge tech dept. I will rollback this part of change, and added the fix in pending task lists

Reply:
Fixed in commits https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1414/commits/7f2a59c8221a2cac5c09efa26025e4661bf0bfde

Reply:
Noticed that `TxContextFieldTag->CallerAddress`, `TxContextFieldTag->CalleeAddress` is the same situation. We can fix it later

Reply:
Hi @ed255 I just have another thought. We better fix it as earlier as possible instead of put it in optimisation. This is not purely a cell optimisation problem, instead more like a bug haha. Because design philosiphy we assume `lo`, `hi` < 128 bits however those 2 address type break the rules.  Besides, when I implementing `beingTx.rs`, `callop.rs`, it need some workaround code just to serve this purpose.  

So I suggest we fix specs part asap.